### PR TITLE
Add stable sampling to jexl filtering

### PR DESCRIPTION
**Pull request by brittanystoroz**
Just having trouble with the dependencies here. Using `sha.js`, I initially had problems where the addon couldn't find the `inherits` module that it required. I whitelisted the `inherits` module in `.jpmignore` which solved that problem. 

But then I got an error about not being able to find the `util` dependency that was required in `inherits`. It didn't get installed automatically with `npm install`, so I tried doing an `npm install --save util` and whitelisting it in `.jpmignore` but that didn't work.

I (temporarily and hackily) worked around needing `util` by editing `sha.js/sha256.js` to require `inherits/inherits_browser` instead, which got me a little further.

My next error was `Buffer is not defined`, when trying to check the filter expression. Buffer is a required dependency of `sha.js`. I also tried whitelisting this dependency in `.jpmignore` but with no success.

### CODE REVIEW

**Comment by brittanystoroz** [comment]
Updates from IRC:

**freaktechnik:** I'm not sure how well the sdk loader handles folder names that end in .js...
**freaktechnik:** I'm also not sure how well that will work, since that module uses Buffers, which are not available in SDK extensions (in the global scope)
**brittanystoroz:** ooh hmmmm. do you know if theres another way to do hashing without a lib like that?
**brittanystoroz:** I also tried using js-sha256 (https://www.npmjs.com/package/js-sha256) but got an error here https://github.com/emn178/js-sha256/blob/master/src/sha256.js#L15 "global is not defined"
**freaktechnik:** you can use webcrypto in a page-worker, or maybe even import them with Cu.importGlobalProperties
**freaktechnik:** though I'm not sure that has the web crypto API, but it might

**Comment by brittanystoroz** [comment]
Updated to import and use `crypto` from `Cu.importGlobalProperties`. It seems to work but I can't promise I have any idea what it's all doing.

File: lib/stableSample.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  for (var i = 0; i < view.byteLength; i += 4) {
+    // Using getUint32 reduces the number of iterations needed (we process 4 bytes each time)
+    var value = view.getUint32(i);
+    // toString(16) will give the hex representation of the number without padding
+    var stringValue = value.toString(16);
+    // We use concatenation and slice for padding
+    var padding = '00000000';
+    var paddedValue = (padding + stringValue).slice(-padding.length);
+    hexCodes.push(paddedValue);
+  }
```
Comment (brittanystoroz) [inline lib/stableSample.js]: I just lifted this from MDN, is there an easier/updated way to do this?

File: lib/stableSample.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ */
+function fractionToKey(frac) {
+    // SHA 256 hashes are 64-digit hexadecimal numbers. The largest possible SHA
+    // 256 hash is 2^256 - 1.
+
+  if (frac < 0 || frac > 1) {
+    throw new Error(`frac must be between 0 and 1 inclusive (got ${frac})`);
+  }
+
+  const mult = Math.pow(2, 256) - 1;
```
Comment (brittanystoroz) [inline lib/stableSample.js]: Is the order of operations on this right?

File: lib/stableSample.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * Map from the range [0, 1] to [0, max(sha256)].
+ * @param  {number} frac A float from 0.0 to 1.0.
+ * @return {string} A string in the sha256 hash space. Will be zero padded to 64
+ *   characters.
+ */
+function fractionToKey(frac) {
+    // SHA 256 hashes are 64-digit hexadecimal numbers. The largest possible SHA
+    // 256 hash is 2^256 - 1.
```
Comment (mythmon) [inline lib/stableSample.js]: Nit: this is indented too much

File: lib/stableSample.js (hunk 2 above)
Comment (mythmon) [inline lib/stableSample.js]: Yes

File: lib/stableSample.js (hunk 3 above)
Comment (mythmon) [inline lib/stableSample.js]: I realize now this was copy/pasted from the one in self-repair. Sorry about that. We should fix it in both places.

File: lib/stableSample.js (hunk 1 above)
Comment (mythmon) [inline lib/stableSample.js]: I don't know of a better way to do this. I think it looks fine.

File: lib/stableSample.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+function fractionToKey(frac) {
+  // SHA 256 hashes are 64-digit hexadecimal numbers. The largest possible SHA
+  // 256 hash is 2^256 - 1.
+
+  if (frac < 0 || frac > 1) {
+    throw new Error(`frac must be between 0 and 1 inclusive (got ${frac})`);
+  }
+
+  const mult = Math.pow(2, 256) - 1;
+  const inDecimal = Math.floor(frac * mult);
```
Comment (mythmon) [inline lib/stableSample.js]: I think this is fine, but I want to double check. Both this and the self-repair version are going to have precision errors in the multiplication. The least significant digits will be wrong, and we'll have big jumps. (like 0.011 straight to 0.032 with nothing in between, but a lot smaller numbers).

I think the precision errors all happen at a scale smaller than we care about. What do you think?

File: lib/stableSample.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+  // Saturate at 2**256 - 1
+  if (hexDigits.length > 64) {
+    hexDigits = Array(65).join('f');
+  }
+
+  return hexDigits;
+}
+
+function hex(buffer) {
```
Comment (mythmon) [inline lib/stableSample.js]: I think something like `bufferToHex` might be a better name for this?

**Comment by mythmon** [comment]
This looks good to me. The CI failures are because we didn't pin the version of JPM, and they changed behavior from 1.0 to 1.1. Bad on them, but oh well. I'll file a PR to pin the versions, and this will be fixed.

**Pull request merged**